rtl/src/main/scala/tensil/mutex/LockPool.scala

```diff
@@ -0,0 +1,145 @@
+  val lockControl = io.lock
+
+  val lock = RegInit(
+    VecInit(
+      Array.fill(numLocks)(
+        zero(new ConditionalReleaseLock(gen, numActors, maxDelay))
+      )
+    )
+  )
+
+  val block = VecInit(for ((a, id) <- actor.zipWithIndex) yield {
+    // block by default
+    io.actor(id).out.noenq()
+    a.ready := false.B
+    val requiredLockId = select(a.bits)
+    val requiredLock   = lock(requiredLockId)
+    val blocked =
+      (requiredLock.held && requiredLock.by =/= id.U) || (lockControl.fire && lockControl.bits.lock === requiredLockId && lockControl.bits.by =/= id.U)
+    when(!blocked) {
+      // allow actor to proceed when not blocked
+      io.actor(id).out <> a
+    }
+    blocked
+  })
+
+  // locked output facilitates testing - not required to be connected in production
+  io.locked.bits <> lockControl.bits
+  io.locked.valid := lockControl.fire
+
+  // for signalling when deadlocked
+  io.deadlocked.noenq()
+
+  def acquire(l: ConditionalReleaseLock[T]): Unit = {
+    when(lockControl.valid) {
+      l.held := lockControl.bits.acquire
+      l.by := lockControl.bits.by
+      l.cond <> lockControl.bits.cond
+    }
+  }
+  def release(l: ConditionalReleaseLock[T]): Unit = {
+    l.held := false.B
+  }
+  // this covers the case where the condition happens to be observed on the same
+  // cycle the lock would have been acquired
+  val incomingObserved =
+    lockControl.valid && actor(lockControl.bits.by).fire && actor(
+      lockControl.bits.by
+    ).bits === lockControl.bits.cond
+  // ready for new lock requests by default
+  val requestedLock = lock(lockControl.bits.lock)
+  // when requested lock held by a different actor than the one requesting, block until lock is released
+  lockControl.ready := !(requestedLock.held && lockControl.bits.by =/= requestedLock.by)
+  for ((l, id) <- lock.zipWithIndex) {
+    val incoming = lockControl.bits.lock === id.U
+    val observed = actor(l.by).fire && actor(l.by).bits === l.cond
+    when(l.held) {
+      when(observed) {
+        when(incoming) {
+          when(incomingObserved) {
+            release(l)
+          }.otherwise {
+            // release then immediately try to acquire
+            release(l)
+            acquire(l)
+          }
+        }.otherwise {
+          release(l)
+        }
+      }.otherwise {
+        when(incoming) {
+          when(l.by === lockControl.bits.by) {
+            // same actor acquiring lock, so allowed
+            acquire(l)
+          }
+        }
+      }
+    }.otherwise {
+      when(incoming) {
+        when(!incomingObserved) {
+          acquire(l)
+        }
+      }
+    }
+  }
+
+  // when all actors are blocked, allow the actor holding lock 0 to proceed
+  when(block.reduce((a, b) => a && b)) {
+    io.actor(lock(0.U).by).out <> actor(lock(0.U).by)
+    // signal out when deadlock happens so we know that something is wrong
+    io.deadlocked.bits := true.B
+    io.deadlocked.valid := true.B
```

Comment:
Is this possible to create the deadlock by supplying specific sequence of instructions?

Reply:
Good question, I think it's possible but tricky to induce because we always request locks at the same time as we actually issue the memory request. Lets do a brainstorm on this out of band.